### PR TITLE
Add ESA CCI land cover dataset

### PR DESCRIPTION
**Pull request by pjhartzell**
### ESA Climate Change Initiative (CCI) Land Cover

**Home:** https://cds.climate.copernicus.eu/cdsapp#!/dataset/satellite-land-cover?tab=overview

The ESA CCI land cover dataset consists of global annual land cover maps at 300m spatial resolution from 1992 to 2020. The source data are NetCDF files that contain five variables: one contains the land cover classes; the remaining four are quality indicators. The source NetCDF files are processed to tiled COGs

### STAC Collections
- [esa-cci-lc](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/esa-cci-lc): COG data
- [esa-cci-lc-netcdf](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/esa-cci-lc-netcdf): source NetCDF data

### PC Test
- [Dataset group page](https://planetarycomputer-test.microsoft.com/dataset/group/esa-cci-lc)
- [Data Explorer view](https://planetarycomputer-test.microsoft.com/explore?c=15.8176%2C-4.5381&z=8.00&v=2&d=esa-cci-lc&m=2017&r=Classification&s=false%3A%3A100%3A%3Atrue&sr=desc&ae=0)

image

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

The 2016 and 2017 data have been ingested into PC Test.

## Checklist:

Please delete options that are not relevant.

- [x] I have performed a self-review
- [ ] Changelog has been updated
- [ ] Documentation has been updated
- [ ] Unit tests pass locally (./scripts/test)
- [x] Code is linted and styled (./scripts/format)

### CODE REVIEW

File: datasets/esa-cci-lc/collection/esa-cci-lc-netcdf/description.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+The ESA Climate Change Initiative (CCI) [Land Cover dataset](https://cds.climate.copernicus.eu/cdsapp#!/dataset/satellite-land-cover?tab=overview) provides consistent global annual land cover maps at 300m spatial resolution from 1992 to 2020. The land cover classes are defined using the United Nations Food and Agriculture Organization's (UN FAO) [Land Cover Classification System](https://www.fao.org/land-water/land/land-governance/land-resources-planning-toolbox/category/details/en/c/1036361/) (LCCS). In addition to the land cover maps, four quality flags are produced to document the reliability of the classification and change detection. 
+
+The data in this Collection are the original NetCDF files accessed from the [Copernicus Climate Data Store](https://cds.climate.copernicus.eu/#!/home).
```
Comment (TomAugspurger) [inline datasets/esa-cci-lc/collection/esa-cci-lc-netcdf/description.md:3]:
```suggestion
The data in this Collection are the original NetCDF files accessed from the [Copernicus Climate Data Store](https://cds.climate.copernicus.eu/#!/home). We recommend users use the [`esa-cci-lc` Collection](planetarycomputer.microsoft.com/dataset/esa-cci-lc), which provides the data as Cloud Optimized GeoTIFFs.
```

File: datasets/esa-cci-lc/collection/esa-cci-lc/description.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (TomAugspurger) [inline datasets/esa-cci-lc/collection/esa-cci-lc/description.md:1]:
```suggestion
```

File: datasets/esa-cci-lc/esa_cci_lc.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                    "https://planetarycomputer.microsoft.com/api/stac/v1/"
+                    "collections/esa-cci-lc-netcdf/items"
+                ),
+            )
+
+            for item in items:
+                id_parts = item.id.split("-")
+                tile = id_parts[-1]
+                version = id_parts[-2]
+                year = id_parts[-3]
```
Comment (TomAugspurger) [inline datasets/esa-cci-lc/esa_cci_lc.py:38]:
```suggestion
                *_, year, version, tile = id_parts
```

(double check I got that right)

File: datasets/esa-cci-lc/esa_cci_lc.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        with TemporaryDirectory() as tmp_dir:
+            storage, nc_path = storage_factory.get_storage_for_file(asset_uri)
+            tmp_nc_path = str(Path(tmp_dir, Path(nc_path).name))
+            storage.download_file(nc_path, tmp_nc_path)
+
+            items = cog_create_items(
+                nc_path=tmp_nc_path,
+                cog_dir=tmp_dir,
+                nc_api_url=(
```
Comment (TomAugspurger) [inline datasets/esa-cci-lc/esa_cci_lc.py:28]: (Making an item locally now to confirm, but writing this first)

I this set as some kind of "derived_from" link? I *think* we have something special in our API for this. I'm pretty sure that if you have a path here like `esa-cci-lc-netcdf/items/{item_id}` (substituting for the real item ID), our STAC API server will dynamically transform that to `https://planetarycomputer.microsoft.com/api/stac/v1/collections/esa-cci-lc-netcdf/items/{item_id}`

Comment (TomAugspurger) [reply datasets/esa-cci-lc/esa_cci_lc.py:28]: I see that it is a derived from: https://github.com/stactools-packages/esa-cci-lc/blob/de196366c565e4e2eff63b63be3a4cb609f54390/src/stactools/esa_cci_lc/cog/stac.py#L122-L131

We have an example of this for sentinel-1-rtc derived from sentinel-1-grd. See https://github.com/microsoft/planetary-computer-tasks/blob/8586b72d6880cd312065d782109e10efa72c8127/datasets/sentinel-1-rtc/s1rtc.py#L82-L93.

So you'd want `collections/esa-cci-lc-netcdf/items/<netcdf-item-id>`.

File: datasets/esa-cci-lc/esa_cci_lc.py (hunk 3 above)
Comment (pjhartzell) [reply datasets/esa-cci-lc/esa_cci_lc.py:38]: Yeah, very nice! Haven't seen this before. "Extended Iterable Unpacking". I like.

File: datasets/esa-cci-lc/esa_cci_lc.py (hunk 4 above)
Comment (pjhartzell) [reply datasets/esa-cci-lc/esa_cci_lc.py]: Sounds good. Updated.

File: datasets/esa-cci-lc/dataset.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  - id: esa-cci-lc-netcdf
+    template: ${{ local.path(./collection/esa-cci-lc-netcdf) }}
+    class: esa_cci_lc:EsaCciLcNetcdf
+    asset_storage:
+      - uri: blob://landcoverdata/esa-cci-lc/
+        chunks:
+          options:
+            name_starts_with: netcdf/
+            extension: [.nc]
+            chunk_length: 2  # for testing; set to 10 for production
```
Comment (TomAugspurger) [inline datasets/esa-cci-lc/dataset.yaml:26]: I'll adjust these as I do the production runs.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**